tensorflow/core/kernels/sparse/sparse_mat_mul_op.cc

```diff
@@ -417,6 +411,136 @@ class CSRSparseMatMulGPUOp : public OpKernel {
+
+    // The SpGEMMDescr contains intermediate results so a separate descr must be
+    // created for each multiply within the batch.
+    std::vector<GpuSparseSpGEMMDescr> gemmDesc(batch_size);
+    std::vector<GpuSparseConstSpMatDescr> matA(batch_size);
+    std::vector<GpuSparseConstSpMatDescr> matB(batch_size);
+    std::vector<GpuSparseSpMatDescr> matC(batch_size);
+    size_t maxBufferSize1 = 0;
+
+    for (int i = 0; i < batch_size; ++i) {
+      OP_REQUIRES_OK(ctx, gemmDesc[i].Initialize());
+      OP_REQUIRES_OK(ctx,
+                     matA[i].InitializeCsr(
+                         a_input_dense_shape(a_input_dense_shape.size() - 2),
+                         a_input_dense_shape(a_input_dense_shape.size() - 1),
+                         a_input_matrix->col_indices_vec(i).size(),
+                         a_input_matrix->row_pointers_vec(i).data(),
+                         a_input_matrix->col_indices_vec(i).data(),
+                         a_input_matrix->values_vec<T>(i).data()));
+      OP_REQUIRES_OK(ctx,
+                     matB[i].InitializeCsr(
+                         b_input_dense_shape(b_input_dense_shape.size() - 2),
+                         b_input_dense_shape(b_input_dense_shape.size() - 1),
+                         b_input_matrix->col_indices_vec(i).size(),
+                         b_input_matrix->row_pointers_vec(i).data(),
+                         b_input_matrix->col_indices_vec(i).data(),
+                         b_input_matrix->values_vec<T>(i).data()));
+      OP_REQUIRES_OK(ctx,
+                     matC[i].InitializeCsr<int, T>(
+                         a_input_dense_shape(a_input_dense_shape.size() - 2),
+                         b_input_dense_shape(b_input_dense_shape.size() - 1), 0,
+                         nullptr, nullptr, nullptr));
+
+      // Determine the maximum size for buffer1 across the batch.
+      // buffer1 will be reused across the batch as each individual gemm will
+      // overwrite it in the second workEstimation call bellow.
+      size_t bufferSize1;
+      OP_REQUIRES_OK(ctx, cudaSparse.SpGEMM_workEstimation<T>(
+                              matA[i], matB[i], matC[i], gemmDesc[i],
+                              &bufferSize1, nullptr));
+      if (bufferSize1 > maxBufferSize1) {
+        maxBufferSize1 = bufferSize1;
+      }
+    }  // End for over batch_size
+
+    Tensor buffer1Tensor;
+    OP_REQUIRES_OK(
+        ctx, ctx->allocate_temp(
+                 DT_INT8, TensorShape({static_cast<int64_t>(maxBufferSize1)}),
+                 &buffer1Tensor));
+    void* buffer1 = buffer1Tensor.flat<int8>().data();
+
+    // Buffer2 is generated along with the number of non-zero elements (NNZ) and
+    // (along with the gemmDesc) contains intermediate products needed for
+    // writing the final output. Because the final outputs cannot be allocated
+    // until the NNZ is computed for the entire batch, we need to keep
+    // batch_size buffer2 temoraries.
```

Comment:
nit: spelling "temporaries"

Reply:
Isn't this a bit unfortunate?  According to the docs, 

> the first invocation of `cusparseSpGEMM_compute` provides an upper bound of the memory required for the computation that is generally several times larger of the actual memory used

So we're potentially allocating many larger-than-necessary buffers, one for each element in the batch.  Would it make sense to allocate a single `buffer2` (that potentially grows as needed), and a potentially growing buffer to store the final columns/values?

Reply:
There are a few options to optimize. I wasn't sure which way to go, so I took the simplest option, which also has the virtue of being the fastest (assuming sufficient memory is available).

If we go with a growing result buffer, we can either reallocate for every element of the batch or estimate the output size based on the NNZ counts observed so far within the batch. A few options ...

* Reallocate at every iteration (e.g. assume all remaining gemms require size zero).
* Reallocate by multiples of two to fit the size required at each iteration
* Assume each remaining GEMM requires the same size as the largest NNZ observed so far in batch.
* Assume each remaining GEMM requires Average NNZ observed so far in batch.
* Either of the above, but add a statistical margin, like an extra 10% or some multiple of the stdev, or space for batch_size+1 outputs.

The "max" option could overestimate the required memory by a lot if one of the gemms is significantly more dense than all the others (worst case by a factor of batch size). Of course, this applies to the "average" case as well, if the outlier happens to be the first gemm. In strange corner cases, this could be larger than the "several times" over allocation in the current implementation, and could persist to downstream ops, unless the output is always resized to an exact fit. Only the "dumbest" first option will never over allocate memory. The second option would bound the over-allocation by a factor of two.

I'm happy to implement any of these. @cantonios, which do you think is best?

Reply:
I was thinking either the 2nd or the 4th one.  I had been debating between those two options myself before your PR when we found out CUDA 12 dropped the old API.  The final output should be resized if it is overly large.

I understand all those tradeoffs, so I'm also on the fence about the best approach.  In the current implementation, the intermediate buffers are all still live when we allocate the result buffers, so the total memory allocated will potentially be considerably more than 2x the final required result size.  It kind of depends how much "several times larger" the intermediate buffers actually are in practice.

I leave it to you to decide what's best here.  I'm okay with the current solution, but thought I would flag my concerns since I had been thinking about it myself over the last week or so.

Reply:
I went with using separate intermediate buffers for each gemm in the batch and then finishing up with a single concat across the outputs.
@cantonios PTAL.